File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();
        int cnt = 0;
        int i = 0, j = 0;
        map<char, int>mp;
        while(i < n)
        {
            if(s[i] == 'a' || s[i] == 'b' || s[i] == 'c') {
                mp[s[i]]++;
            }

            if(mp.size() == 3)
            {
                while(mp.size() == 3)
                {
                    cnt += (n - i);
                    mp[s[j]]--;
                    if(mp[s[j]] == 0) mp.erase(s[j]);
                    j++;
                }
            }
            i++;
        }
        return cnt;
    }
};

File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        bad = 0
        total = n * (n + 1) // 2
        j = 0

        def default_value():
            return 0
            
        mp = defaultdict(default_value)

        for i in range(n):
            if s[i] in mp:
                mp[s[i]] += 1
            else:
                mp[s[i]] = 1
            
            if mp['a'] > 0 and mp['b'] > 0 and mp['c'] > 0:
                while mp['a'] > 0 and mp['b'] > 0 and mp['c'] > 0:
                    mp[s[j]]-=1
                    j+=1

            bad += (i - j + 1)

        return total - bad